rfcs/framework-0072-pluggable-transfers.md

```diff
@@ -0,0 +1,162 @@
+a direct access to the referenced data.
+
+During "dumping" the file record, a links section with `content` uri is created. 
+This is a bit confusing as clients do not know which URL should be used.
+
+Proposal: keep the uri only inside the initial file record. Do not serialize it
+afterwards. Clients should use the links section for all the URIs. The content
+url can return a HTTP 302 redirect to the actual file location. This way counters
+will work as well.
+
+- [ ] FETCH transfer type and private URIs
+
+The FETCH transfer type is used to fetch the file from a remote location. The uri
+of the external location is visible to anyone who has a read access to the record,
+not just the uploader. As the uri might contain credentials (e.g. S3 signed URL),
+this might be a security issue.
+
+Proposal: do not serialize the uri of the FETCH transfer type. the fetch_file task
+would have to be changed to not to use the service but fetch the uri directly.
+
+- [ ] IfFileIsLocal generators
+
+The permissions in https://github.com/inveniosoftware/invenio-records-resources/blob/master/invenio_records_resources/services/files/generators.py#L27
+take into account only the files with storage class (== transfer type) "L". This means that
+the permissions will fail for "M" (Multipart transport) files. Is this intended?
+
+The same problem is in invenio-rdm-records/services/generators.py
+
+Proposal: either change the meaning of "Locality" and keep the permissions as they are
+or change the permissions to take into account all the files. Example from RDM that I do 
+not understand (can_draft_set_content_files - why there is the can_review permission and 
+not record owner?):
+
+```python
+    can_draft_create_files = can_review
+    can_draft_set_content_files = [
+        # review is the same as create_files
+        IfFileIsLocal(then_=can_review, else_=[SystemProcess()])
+    ]
+    can_draft_get_content_files = [
+        # preview is same as read_files
+        IfFileIsLocal(then_=can_draft_read_files, else_=[SystemProcess()])
+    ]
+    can_draft_commit_files = [
+        # review is the same as create_files
+        IfFileIsLocal(then_=can_review, else_=[SystemProcess()])
+    ]
+```
+
+- [ ] MultipartStorageExt
+
+In order to make direct multipart upload to remote storage working we need a PyFS storage
+to implement multipart aware methods. These are currently summarized in the MultipartStorageExt
+interface that lives in invenio-records-resources - 
+[MultipartStorageExt class in pull request](https://github.com/mesemus/invenio-records-resources/pull/3/files#diff-ab7d7ef448eaae0f109622e167d9168d7f0ad56dbba383d6d681b9cc17b62bcfR14)
+
+
+We can not directly implement this interface in storage backends (for example invenio-s3) as
+it would introduce a dependency in the wrong direction.
+
+Proposal: either move the MultipartStorageExt to invenio-files-rest/storage/pyfs or do not check
```

Comment:
The semantics/naming of `MultipartStorageExt` look very much bound to the `init/content/commit` terminology, so they would look somewhat out of place in `invenio-files-rest`. Thus I'm leaning more towards keeping a "Protocol"-style coupling.

Another solution, might be to move the relevant `invenio-s3` parts that implement the `MultipartStorageExt` interface in this module under "contrib" or similar and do some `try/except` around imports to do feature detection.

Reply:
Ok, let's keep the protocol style, I've made it more readable in the code.